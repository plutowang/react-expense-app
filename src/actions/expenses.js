import database from "../firebase/firebase";

// Expense Action Generators

// ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = expenseData => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createAt = 0
    } = expenseData || {};
    const expense = {
      description,
      note,
      amount,
      createAt
    };
    // By returning the `promise chain` we can continue chaining on over here
    return database
      .ref("expenses")
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

// REMOVE_ALL_EXPENSE
export const removeAllExpense = () => ({
  type: "REMOVE_ALL_EXPENSE"
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});
