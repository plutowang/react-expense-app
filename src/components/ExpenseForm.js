import React from "react";
import moment from "moment";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";
import { SingleDatePicker } from "react-dates";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createAt: props.expense ? moment(props.expense.createAt) : moment(),
      calendarFocused: false,
      error: "",
      isUpdate: props.isUpdate
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = e => {
    const amount = e.target.value ? e.target.value : "";
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = createAt => {
    if (createAt) {
      this.setState(() => ({ createAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      const error = "Please provide description and amount";
      this.setState(() => ({ error }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        note: this.state.note,
        createAt: this.state.createAt.valueOf()
      });
    }
  };

  render() {
    return (
      <div>
        {!!this.state.error && <h3>{this.state.error}</h3>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            name="description"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            name="amount"
            placeholder="$"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            name="note"
            id="note"
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>
            {this.state.isUpdate ? "Update Expense" : "Add Expense"}
          </button>
        </form>
      </div>
    );
  }
}
