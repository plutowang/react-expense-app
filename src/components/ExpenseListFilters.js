import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      calendarFocused: null
    };
  }
  onInputChange = e => {
    this.props.setTextFilter(e.target.value);
  };
  onSelectChange = e => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    }
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };
  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  render() {
    return (
      <div>
        <input
          type="text"
          name="textFilter"
          defaultValue={this.props.filters.text}
          placeholder="text filter"
          onChange={this.onInputChange}
        />
        <select
          name="sortBy"
          id="sortBy"
          defaultValue={this.props.filters.sortBy}
          onChange={this.onSelectChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId="startDateId"
          endDate={this.props.filters.endDate}
          endDateId="endDateId"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
