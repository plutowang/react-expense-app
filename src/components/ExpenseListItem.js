import React from "react";
import { Link } from "react-router-dom";

// Export a stateless function components
// show up description, amount, createdAt
export const ExpenseListItem = ({ id, description, amount, createAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      ${amount} -- {createAt}
    </p>
  </div>
);

export default ExpenseListItem;
