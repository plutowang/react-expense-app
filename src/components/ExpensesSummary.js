import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import numeral from "numeral";
import selectExpensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
  const total = numeral(expensesTotal / 100).format("$0,0.00");
  const expr = expensesCount > 1 ? "expenses" : "expense";
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expensesCount}</span> {expr} totalling <span>{total}</span>
        </h1>
        <div className="page-header__action">
        <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  const expenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesCount: expenses.length,
    expensesTotal: selectExpensesTotal(expenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
