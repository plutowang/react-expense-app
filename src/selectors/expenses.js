import moment from "moment";
// Get Visiable Expenses

const getVisableExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createAtMoment = moment(expense.createAt);
      // if all three of these are true,
      // we have complete mathch and
      // we want to return `True` form the filter
      // undifined startDate or createAt is after the startDate
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createAtMoment, "day")
        : true;
      const textMatch =
        typeof text !== "string" ||
        (expense.description &&
          expense.description.toLowerCase().includes(text.toLowerCase())) ||
        (expense.note &&
          expense.note.toLowerCase().includes(text.toLowerCase()));

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      // sort by deceending
      if (sortBy === "date") {
        return a.createAt < b.createAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return b.amount - a.amount;
      }
    });
};

export default getVisableExpenses;
