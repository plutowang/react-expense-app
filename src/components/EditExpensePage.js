import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpenses, startRemoveExpenses } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startEditExpenses(this.props.expense.id, expense);
    this.props.history.push("/");
  };
  onRemove = () => {
    this.props.startRemoveExpenses();
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Update Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            isUpdate={true}
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <div className="form">
            <button className="button button--remove" onClick={this.onRemove}>
              Remove Expense
            </button>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});
const mapDispatchToProps = (dispatch, props) => ({
  startEditExpenses: (id, expense) => dispatch(startEditExpenses(id, expense)),
  startRemoveExpenses: () =>
    dispatch(startRemoveExpenses({ id: props.match.params.id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
