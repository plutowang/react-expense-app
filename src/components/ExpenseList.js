import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import { startRemoveAllExpenses } from "../actions/expenses";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => (
  <div>
    <h1>Expense List</h1>
    <button onClick={props.removeAll}>
      Remove All
    </button>
    {props.expenses && props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(expense => (
        <ExpenseListItem key={expense.id} {...expense} />
      ))
    )}
  </div>
);
const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};
const mapDispatchToProps = dispatch => ({
  removeAll: ()=>dispatch(startRemoveAllExpenses())
});
// HOC
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseList);
