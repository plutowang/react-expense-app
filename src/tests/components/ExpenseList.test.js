import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("should render ExpenseList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseList empty expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});


test("should remove all expenses", () => {
  const removeAllExpenses = jest.fn();
  const wrapper = shallow(<ExpenseList expenses={expenses} removeAll={removeAllExpenses}/>);
  wrapper.find("button").simulate("click")
  expect(removeAllExpenses).toHaveBeenLastCalledWith()
});