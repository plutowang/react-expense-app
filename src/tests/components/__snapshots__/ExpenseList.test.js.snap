// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList empty expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={54500}
    createAt={0}
    description="January Rent"
    id="0"
    key="0"
    note="This was the fial payment for that address"
  />
  <ExpenseListItem
    amount={900}
    createAt={-345600000}
    description="Gum Bill"
    id="1"
    key="1"
    note="This was the fial payment for gum"
  />
  <ExpenseListItem
    amount={1200}
    createAt={345600000}
    description="Water Bill"
    id="2"
    key="2"
    note="This was the fial payment for water"
  />
  <ExpenseListItem
    amount={890}
    createAt={691200000}
    description="Gym Bill"
    id="3"
    key="3"
    note="This was the fial payment for gym"
  />
</div>
`;
